src/shuffler-codereview.c

```diff
@@ -0,0 +1,1228 @@
+            nextIndex++;
+        }
+        // shuffle the remaining indicies.
+        for (int i = filteredPoolSize - 1; i > 0; i--) {
+            int r = tinymt32_generate_uint32(&masterSeed) % (i + 1);
+            int tmp;
+            SWAP(filteredIndicies[i], filteredIndicies[r], tmp);
+        }
+        // fill the first 3 party slots with pokemon from pool1.
+        for (int i = 0; i < trainer->partySize - 1; i++) {
+            partyIndicies[i] = filteredIndicies[i];
+        }
+
+        // now we do the same for pool2, except we're only picking 1.
+        filteredPoolSize = tt->pool2Size;
+        nextIndex = 0;
+        // filter out possible pokemon that are of the wrong level.
+        for (int i = 0; i < tt->pool2Size; i++) {
+            if (!PokemonIsValidLevel(&tt->pool2[i], level)) {
+                filteredPoolSize--;
+                continue;
+            }
+            filteredIndicies[nextIndex] = i;
+            nextIndex++;
+        }
+        // shuffle the remaining indicies.
+        for (int i = filteredPoolSize - 1; i > 0; i--) {
+            int r = tinymt32_generate_uint32(&masterSeed) % (i + 1);
+            int tmp;
+            SWAP(filteredIndicies[i], filteredIndicies[r], tmp);
+        }
+        // put a pokemon from pool2 into the final party slot.
+        partyIndicies[trainer->partySize - 1] = filteredIndicies[0] + tt->pool1Size;
+
+        // if there is no pool3, we're done.
+        if (!tt->pool3Size) {
+            break;
+        }
+        // we're only 1/3 to switch a pokemon with one from pool3. if we lose a
+        // 1/3 roll, we're done.
+        int r = tinymt32_generate_uint32(&masterSeed) % 3;
+        if (r != 2) {
+            break;
+        }
+
+        // if there is a pool3, we need to replace one of the chosen pokemon
+        // from pool1 with one pokemon from pool3.
+        filteredPoolSize = tt->pool3Size;
+        nextIndex = 0;
+        // filter out possible pokemon that are of the wrong level.
+        for (int i = 0; i < tt->pool2Size; i++) {
+            if (!PokemonIsValidLevel(&tt->pool3[i], level)) {
+                filteredPoolSize--;
+                continue;
+            }
+            filteredIndicies[nextIndex] = i;
+            nextIndex++;
+        }
+        // it's possible we end up with no possible pokemon here. if that's true, we're done.
+        if (!filteredPoolSize) {
+            break;
+        }
+
+        // shuffle the remaining indicies.
+        for (int i = filteredPoolSize - 1; i > 0; i--) {
+            int r = tinymt32_generate_uint32(&masterSeed) % (i + 1);
+            int tmp;
+            SWAP(filteredIndicies[i], filteredIndicies[r], tmp);
+        }
+        // put a pokemon from pool3 into the first party slot, overwriting a
+        // pokemon chosen from pool1.
+        partyIndicies[0] = filteredIndicies[0] + tt->pool1Size + tt->pool2Size;
+
+        break;
+    }
+    case TB_STRAT_GYM_LEADER: {
```

Comment:
This code looks so much like the TB_STRAT_GYM_MEMBER case, that I'm suspicious. Either comment, or is it possible to make all of this less duplicated?